# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' sp_dist_euclidean_cpp
#' This function calculates Euclidean distance between cell locations and spot locations.
#' @author Qi Gao
#' @param cloc matrix of location coordinates of the cells. Each row represents a cell.
#' @param sloc matrix of location coordinates of the spots. Each row represents a spot.
#' @return matrix of the Euclidean distance, number of cells (row) by number of spots (column) 
#' @export
sp_dist_euclidean_cpp <- function(cloc, sloc) {
    .Call(`_GeneScape_sp_dist_euclidean_cpp`, cloc, sloc)
}

#' cumsum_cpp
#' This function calculates the cumulative sum (partial sum) of a vector.
#' @author Qi Gao
#' @param x vector to be summed.
#' @return vector of the cumulative sum (partial sum)
#' @export
cumsum_cpp <- function(x) {
    .Call(`_GeneScape_cumsum_cpp`, x)
}

#' sample_int
#' This function samples integers from 1 to an upper limit integer value with equal probability without replacement.
#' @author Qi Gao
#' @param maxvalue the max integer to be sampled.
#' @param nsample the number of integer to be sampled.
#' @return vector of the integers in the sample
#' @export
sample_int <- function(maxvalue, nsample) {
    .Call(`_GeneScape_sample_int`, maxvalue, nsample)
}

#' downSampleRead
#' This function uses down-sampling to obtain the reads in each spot. 
#' @author Qi Gao
#' @param count gene (row) by cell (column) matrix. Gene read count in each cell.
#' @param nread cell (row) by spot (column) matrix. Target number of reads sampled from each cell in each spot.
#' @return gene (row) by spot (column) matrix. Gene expression level in each spot.
#' @export
downSampleRead <- function(count, nread) {
    .Call(`_GeneScape_downSampleRead`, count, nread)
}

#' knnassign
#' This function uses KNN sampling to simulate gradually changed tissue types.
#' @author Qi Gao
#' @param distmat Spot distance matrix 
#' @param ttype Input tissue type of each spot
#' @param nttype total number of possible tissue types. should be no less than the max value of ttype and the 
#' number of unique values in ttype. some tissue types may not exist in ttype
#' @param k parameter for KNN sampling
#' @return List of tissue type weight and updated tissue type with gradual change
#' @export
knnassign <- function(distmat, ttype, nttype, k) {
    .Call(`_GeneScape_knnassign`, distmat, ttype, nttype, k)
}

#' changeLibSize
#' This function modifies the simulated counts so that the total 
#' read count in each spot is the same as the example data.
#' @author Qi Gao
#' @param count simulated matrix of read count, rows as genes and columns as spots
#' @param nreadspot number of total read count in each spot in the example data
#' @param weight a matrix of probability weights for the read count of each gene in each cell being changed.
#' @return a matrix of the updated read count
#' @export
changeLibSize <- function(count, nreadspot, weight) {
    .Call(`_GeneScape_changeLibSize`, count, nreadspot, weight)
}

#' estimate_gamma
#' This function estimates gamma distribution shape and rate parameters.
#' @author Qi Gao
#' @param x target data vector
#' @return list of gamma distribution shape and rate parameters
#' @reference Zhi-Sheng Ye, Nan Chen (2017) Closed-Form Estimators for the Gamma Distribution Derived From Likelihood Equations, 
#' The American Statistician, 71:2, 177-181, DOI: 10.1080/00031305.2016.1209129
#' @export
#' 
estimate_gamma <- function(x) {
    .Call(`_GeneScape_estimate_gamma`, x)
}

#' estimate_lognormal
#' This function estimates estimate_lognormal distribution loc and scale parameters.
#' @author Qi Gao
#' @param x target data vector
#' @return list of lognormal distribution loc and scale parameters
#' @reference Ginos, Brenda Faith, "Parameter Estimation for the Lognormal Distribution" (2009). 
#' Theses and Dissertations. 1928. https://scholarsarchive.byu.edu/etd/1928
#' @export
#' 
estimate_lognormal <- function(x) {
    .Call(`_GeneScape_estimate_lognormal`, x)
}

